{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Process all the data using TPP"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [],
   "source": [
    "# TODO: Delete this cell before release\n",
    "# #Change the names of all the fastq files to SRA numbers for testing\n",
    "# from shutil import copyfile\n",
    "# import pandas as pd\n",
    "\n",
    "# samp_df = pd.read_csv('input/Sample_Labels.csv')\n",
    "# for uid, sra in zip(samp_df['Run_UID'],samp_df['SRA Accession']):\n",
    "#     f1 = 'input/'+uid + '_1.fastq'\n",
    "#     f1_out = 'input/'+sra + '_1.fastq'\n",
    "#     f2 = 'input/'+uid + '_2.fastq'\n",
    "#     f2_out = 'input/'+sra + '_2.fastq'\n",
    "    \n",
    "#     copyfile(f1, f1_out, follow_symlinks=False)\n",
    "#     copyfile(f2, f2_out, follow_symlinks=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "******** Run 0: input/fastq/SRR23906893_1.fastq ********\n",
      "******** Run 0: input/fastq/SRR23906897_1.fastq ********\n",
      "# title: Tn-Seq Pre-Processor\n",
      "# date: 03/31/2023 09:59:12\n",
      "# command: python /home/will/src_and_bin/Other_Software/miniconda3/envs/tnseq_avium_abx_fordistro/bin/tpp -himar1 -bwa /home/will/src_and_bin/Other_Software/miniconda3/envs/tnseq_avium_abx_fordistro/bin/bwa -bwa-alg aln -ref input/Genome/GCF_000195955.2_ASM19595v2_genomic.fna -replicon-ids NC_000962.3 -reads1 input/fastq/SRR23906893_1.fastq -reads2 input/fastq/SRR23906893_2.fastq -window-size 6 -primer AACCTGTTA -mismatches 2 -output output/tpp_out/SRR23906893\n",
      "# transposon type: Himar1\n",
      "# protocol type: Sassetti\n",
      "# bwa flags:\n",
      "# read1: input/fastq/SRR23906893_1.fastq\n",
      "# read2: input/fastq/SRR23906893_2.fastq\n",
      "# ref_genome: input/Genome/GCF_000195955.2_ASM19595v2_genomic.fna\n",
      "# replicon_ids: NC_000962.3\n",
      "# total_reads (or read pairs): 2783134\n",
      "# trimmed_reads (reads with valid Tn prefix, and insert size>20bp): 2661285\n",
      "# reads1_mapped: 2581237\n",
      "# reads2_mapped: 2451071\n",
      "# mapped_reads (both R1 and R2 map into genome, and R2 has a proper barcode): 1982480\n",
      "# read_count (TA sites only, for Himar1): 1976507\n",
      "# template_count: 1897309\n",
      "# template_ratio (reads per template): 1.04\n",
      "# TA_sites: 74604\n",
      "# TAs_hit: 44667\n",
      "# density: 0.599\n",
      "# max_count (among templates): 5699\n",
      "# max_site (coordinate): 212278\n",
      "# NZ_mean (among templates): 42.5\n",
      "# FR_corr (Fwd templates vs. Rev templates): 0.748\n",
      "# BC_corr (reads vs. templates, summed over both strands): 1.000\n",
      "# primer_matches: 35 reads (0.0%) contain CTAGAGGGCCCAATTCGCCCTATAGTGAGT (Himar1)\n",
      "# vector_matches: 2046 reads (0.1%) contain CTAGACCGTCCAGTCTGGCAGGCCGGAAAC (phiMycoMarT7)\n",
      "# adapter_matches: 81947 reads (2.9%) contain GATCGGAAGAGCACACGTCTGAACTCCAGTCAC (Illumina/TruSeq index)\n",
      "# misprimed_reads: 26446 reads (1.0%) contain Himar1 prefix but don't end in TGTTA\n",
      "# read_length: 125 bp\n",
      "# mean_R1_genomic_length: 90.4 bp\n",
      "# mean_R2_genomic_length: 72.4 bp\n",
      "# title: Tn-Seq Pre-Processor\n",
      "# date: 03/31/2023 10:07:38\n",
      "# command: python /home/will/src_and_bin/Other_Software/miniconda3/envs/tnseq_avium_abx_fordistro/bin/tpp -himar1 -bwa /home/will/src_and_bin/Other_Software/miniconda3/envs/tnseq_avium_abx_fordistro/bin/bwa -bwa-alg aln -ref input/Genome/GCF_000195955.2_ASM19595v2_genomic.fna -replicon-ids NC_000962.3 -reads1 input/fastq/SRR23906897_1.fastq -reads2 input/fastq/SRR23906897_2.fastq -window-size 6 -primer AACCTGTTA -mismatches 2 -output output/tpp_out/SRR23906897\n",
      "# transposon type: Himar1\n",
      "# protocol type: Sassetti\n",
      "# bwa flags:\n",
      "# read1: input/fastq/SRR23906897_1.fastq\n",
      "# read2: input/fastq/SRR23906897_2.fastq\n",
      "# ref_genome: input/Genome/GCF_000195955.2_ASM19595v2_genomic.fna\n",
      "# replicon_ids: NC_000962.3\n",
      "# total_reads (or read pairs): 5972001\n",
      "# trimmed_reads (reads with valid Tn prefix, and insert size>20bp): 4886129\n",
      "# reads1_mapped: 4423156\n",
      "# reads2_mapped: 3973440\n",
      "# mapped_reads (both R1 and R2 map into genome, and R2 has a proper barcode): 3362136\n",
      "# read_count (TA sites only, for Himar1): 3338336\n",
      "# template_count: 3046366\n",
      "# template_ratio (reads per template): 1.10\n",
      "# TA_sites: 74604\n",
      "# TAs_hit: 47301\n",
      "# density: 0.634\n",
      "# max_count (among templates): 12183\n",
      "# max_site (coordinate): 912723\n",
      "# NZ_mean (among templates): 64.4\n",
      "# FR_corr (Fwd templates vs. Rev templates): 0.805\n",
      "# BC_corr (reads vs. templates, summed over both strands): 1.000\n",
      "# primer_matches: 61 reads (0.0%) contain CTAGAGGGCCCAATTCGCCCTATAGTGAGT (Himar1)\n",
      "# vector_matches: 158 reads (0.0%) contain CTAGACCGTCCAGTCTGGCAGGCCGGAAAC (phiMycoMarT7)\n",
      "# adapter_matches: 104015 reads (1.7%) contain GATCGGAAGAGCACACGTCTGAACTCCAGTCAC (Illumina/TruSeq index)\n",
      "# misprimed_reads: 163962 reads (2.7%) contain Himar1 prefix but don't end in TGTTA\n",
      "# read_length: 125 bp\n",
      "# mean_R1_genomic_length: 92.2 bp\n",
      "# mean_R2_genomic_length: 68.5 bp\n",
      "Creating prefixes file with all prefixes from all runs...\n",
      "Created 'output/tpp_out/fastq_prefixes.txt'.\n",
      "\n",
      "Creating CSV file with all samples processed by TPP...\n",
      "[+] scripts/wig_gb_to_csv.py starting with parameters:\n",
      "....Wiggle prefix(es)         output/tpp_out/SRR23906893, output/tpp_out/SRR23906897\n",
      "....Record ID(s)              NC_000962.3\n",
      "....GenBank file:             input/Genome/GCF_000195955.2_ASM19595v2_genomic.gbff\n",
      "....Unique identifier fields: locus_tag, record_id_start_end_strand\n",
      "....Fields:                   product, regulatory_class, bound_moiety\n",
      "....Output file:              output/fastq.csv\n",
      "[+] Combining Wiggles with prefix '['output/tpp_out/SRR23906893', 'output/tpp_out/SRR23906897']' with Record ID 'NC_000962.3' from 'input/Genome/GCF_000195955.2_ASM19595v2_genomic.gbff'...\n",
      "[+] Finished processing Record ID 'NC_000962.3' successfully in 4.611685 seconds.\n",
      "[+] Finished processing all records successfully in 4.611704 seconds.\n",
      "[+] Writing CSV files...\n",
      "[+] 'output/fastq.csv' written.\n",
      "[+] Success. Exiting...\n",
      "Created 'output/fastq.csv'.\n",
      "\n",
      "********** TPP driver script finished in a total of 977 seconds **********\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "[tn_preprocess] running pre-processing on input/fastq/SRR23906893_1.fastq\n",
      "[tn_preprocess] protocol: Sassetti\n",
      "[tn_preprocess] transposon type: Himar1\n",
      "[tn_preprocess] running pre-processing on input/fastq/SRR23906897_1.fastq\n",
      "[tn_preprocess] protocol: Sassetti\n",
      "[tn_preprocess] transposon type: Himar1\n",
      "[tn_preprocess] One reference genome specified: input/Genome/GCF_000195955.2_ASM19595v2_genomic.fna, containing 1 records.\n",
      "[tn_preprocess] extracting reads...\n",
      "[tn_preprocess] fastq2reads: input/fastq/SRR23906893_1.fastq -> output/tpp_out/SRR23906893.reads1\n",
      "[tn_preprocess] One reference genome specified: input/Genome/GCF_000195955.2_ASM19595v2_genomic.fna, containing 1 records.\n",
      "[tn_preprocess] extracting reads...\n",
      "[tn_preprocess] fastq2reads: input/fastq/SRR23906897_1.fastq -> output/tpp_out/SRR23906897.reads1\n",
      "[tn_preprocess] 1000000 reads processed\n",
      "[tn_preprocess] 1000000 reads processed\n",
      "[tn_preprocess] 2000000 reads processed\n",
      "[tn_preprocess] 2000000 reads processed\n",
      "[tn_preprocess] 3000000 reads processed\n",
      "[tn_preprocess] 4000000 reads processed\n",
      "[tn_preprocess] fastq2reads: input/fastq/SRR23906893_2.fastq -> output/tpp_out/SRR23906893.reads2\n",
      "[tn_preprocess] 5000000 reads processed\n",
      "[tn_preprocess] fastq2reads: input/fastq/SRR23906897_2.fastq -> output/tpp_out/SRR23906897.reads2\n",
      "[tn_preprocess] 1000000 reads processed\n",
      "[tn_preprocess] 1000000 reads processed\n",
      "[tn_preprocess] 2000000 reads processed\n",
      "[tn_preprocess] 2000000 reads processed\n",
      "[tn_preprocess] 3000000 reads processed\n",
      "[tn_preprocess] fixing headers of paired reads for bwa...\n",
      "[tn_preprocess] 4000000 reads processed\n",
      "[tn_preprocess] 5000000 reads processed\n",
      "[tn_preprocess] 1000000 reads processed\n",
      "[tn_preprocess] 2000000 reads processed\n",
      "[tn_preprocess] extracting barcodes and genomic parts of reads...\n",
      "[tn_preprocess] creating output/tpp_out/SRR23906893.trimmed1\n",
      "[tn_preprocess] prefix sequence: AACCTGTTA\n",
      "[tn_preprocess] Looking for start of Tn prefix with P,Q = 16,22 (origin = 19, window size = 6)\n",
      "[tn_preprocess] fixing headers of paired reads for bwa...\n",
      "[tn_preprocess] 1000000 reads processed\n",
      "[tn_preprocess] 2000000 reads processed\n",
      "[tn_preprocess] 1000000 reads processed\n",
      "[tn_preprocess] 3000000 reads processed\n",
      "[tn_preprocess] 4000000 reads processed\n",
      "[tn_preprocess] 5000000 reads processed\n",
      "[tn_preprocess] extracting barcodes and genomic parts of reads...\n",
      "[tn_preprocess] creating output/tpp_out/SRR23906897.trimmed1\n",
      "[tn_preprocess] prefix sequence: AACCTGTTA\n",
      "[tn_preprocess] Looking for start of Tn prefix with P,Q = 16,22 (origin = 19, window size = 6)\n",
      "[tn_preprocess] 2000000 reads processed\n",
      "[tn_preprocess] creating output/tpp_out/SRR23906893.trimmed2\n",
      "[tn_preprocess] creating output/tpp_out/SRR23906893.barcodes2\n",
      "[tn_preprocess] creating output/tpp_out/SRR23906893.genomic2\n",
      "[tn_preprocess] 1000000 reads processed\n",
      "[tn_preprocess] 2000000 reads processed\n",
      "[tn_preprocess] 1000000 reads processed\n",
      "[tn_preprocess] 3000000 reads processed\n",
      "[tn_preprocess] 2000000 reads processed\n",
      "[tn_preprocess] creating output/tpp_out/SRR23906893.barcodes1\n",
      "[tn_preprocess] 4000000 reads processed\n",
      "[tn_preprocess] mapping reads using BWA...(this takes a couple of minutes)\n",
      "[tn_preprocess] /home/will/src_and_bin/Other_Software/miniconda3/envs/tnseq_avium_abx_fordistro/bin/bwa aln input/Genome/GCF_000195955.2_ASM19595v2_genomic.fna output/tpp_out/SRR23906893.trimmed1 > output/tpp_out/SRR23906893.sai1\n",
      "[tn_preprocess] 5000000 reads processed\n",
      "[tn_preprocess] creating output/tpp_out/SRR23906897.trimmed2\n",
      "[tn_preprocess] creating output/tpp_out/SRR23906897.barcodes2\n",
      "[tn_preprocess] creating output/tpp_out/SRR23906897.genomic2\n",
      "[bwa_aln] 17bp reads: max_diff = 2\n",
      "[bwa_aln] 38bp reads: max_diff = 3\n",
      "[bwa_aln] 64bp reads: max_diff = 4\n",
      "[bwa_aln] 93bp reads: max_diff = 5\n",
      "[bwa_aln] 124bp reads: max_diff = 6\n",
      "[bwa_aln] 157bp reads: max_diff = 7\n",
      "[bwa_aln] 190bp reads: max_diff = 8\n",
      "[bwa_aln] 225bp reads: max_diff = 9\n",
      "[bwa_aln_core] calculate SA coordinate... 9.04 sec\n",
      "[bwa_aln_core] write to the disk... 0.02 sec\n",
      "[bwa_aln_core] 262144 sequences have been processed.\n",
      "[bwa_aln_core] calculate SA coordinate... 9.51 sec\n",
      "[bwa_aln_core] write to the disk... 0.02 sec\n",
      "[bwa_aln_core] 524288 sequences have been processed.\n",
      "[bwa_aln_core] calculate SA coordinate... 8.50 sec\n",
      "[bwa_aln_core] write to the disk... 0.02 sec\n",
      "[bwa_aln_core] 786432 sequences have been processed.\n",
      "[bwa_aln_core] calculate SA coordinate... 8.26 sec\n",
      "[bwa_aln_core] write to the disk... 0.02 sec\n",
      "[bwa_aln_core] 1048576 sequences have been processed.\n",
      "[bwa_aln_core] calculate SA coordinate... 8.06 sec\n",
      "[bwa_aln_core] write to the disk... 0.02 sec\n",
      "[bwa_aln_core] 1310720 sequences have been processed.\n",
      "[bwa_aln_core] calculate SA coordinate... 7.93 sec\n",
      "[bwa_aln_core] write to the disk... 0.02 sec\n",
      "[bwa_aln_core] 1572864 sequences have been processed.\n",
      "[bwa_aln_core] calculate SA coordinate... 8.63 sec\n",
      "[bwa_aln_core] write to the disk... 0.02 sec\n",
      "[bwa_aln_core] 1835008 sequences have been processed.\n",
      "[bwa_aln_core] calculate SA coordinate... 7.96 sec\n",
      "[bwa_aln_core] write to the disk... 0.02 sec\n",
      "[bwa_aln_core] 2097152 sequences have been processed.\n",
      "[bwa_aln_core] calculate SA coordinate... 8.41 sec\n",
      "[bwa_aln_core] write to the disk... 0.02 sec\n",
      "[bwa_aln_core] 2359296 sequences have been processed.\n",
      "[bwa_aln_core] calculate SA coordinate... 8.53 sec\n",
      "[bwa_aln_core] write to the disk... 0.02 sec\n",
      "[bwa_aln_core] 2621440 sequences have been processed.\n",
      "[bwa_aln_core] calculate SA coordinate... 1.21 sec\n",
      "[bwa_aln_core] write to the disk... 0.00 sec\n",
      "[bwa_aln_core] 2661285 sequences have been processed.\n",
      "[main] Version: 0.7.17-r1188\n",
      "[main] CMD: /home/will/src_and_bin/Other_Software/miniconda3/envs/tnseq_avium_abx_fordistro/bin/bwa aln input/Genome/GCF_000195955.2_ASM19595v2_genomic.fna output/tpp_out/SRR23906893.trimmed1\n",
      "[main] Real time: 88.063 sec; CPU: 87.920 sec\n",
      "\n",
      "[tn_preprocess] /home/will/src_and_bin/Other_Software/miniconda3/envs/tnseq_avium_abx_fordistro/bin/bwa aln input/Genome/GCF_000195955.2_ASM19595v2_genomic.fna output/tpp_out/SRR23906893.genomic2 > output/tpp_out/SRR23906893.sai2\n",
      "[tn_preprocess] 1000000 reads processed\n",
      "[bwa_aln] 17bp reads: max_diff = 2\n",
      "[bwa_aln] 38bp reads: max_diff = 3\n",
      "[bwa_aln] 64bp reads: max_diff = 4\n",
      "[bwa_aln] 93bp reads: max_diff = 5\n",
      "[bwa_aln] 124bp reads: max_diff = 6\n",
      "[bwa_aln] 157bp reads: max_diff = 7\n",
      "[bwa_aln] 190bp reads: max_diff = 8\n",
      "[bwa_aln] 225bp reads: max_diff = 9\n",
      "[bwa_aln_core] calculate SA coordinate... 7.42 sec\n",
      "[bwa_aln_core] write to the disk... 0.02 sec\n",
      "[bwa_aln_core] 262144 sequences have been processed.\n",
      "[bwa_aln_core] calculate SA coordinate... 7.53 sec\n",
      "[bwa_aln_core] write to the disk... 0.02 sec\n",
      "[bwa_aln_core] 524288 sequences have been processed.\n",
      "[bwa_aln_core] calculate SA coordinate... 7.52 sec\n",
      "[bwa_aln_core] write to the disk... 0.02 sec\n",
      "[bwa_aln_core] 786432 sequences have been processed.\n",
      "[bwa_aln_core] calculate SA coordinate... 7.54 sec\n",
      "[bwa_aln_core] write to the disk... 0.02 sec\n",
      "[bwa_aln_core] 1048576 sequences have been processed.\n",
      "[bwa_aln_core] calculate SA coordinate... 7.52 sec\n",
      "[bwa_aln_core] write to the disk... 0.02 sec\n",
      "[bwa_aln_core] 1310720 sequences have been processed.\n",
      "[bwa_aln_core] calculate SA coordinate... 7.46 sec\n",
      "[bwa_aln_core] write to the disk... 0.02 sec\n",
      "[bwa_aln_core] 1572864 sequences have been processed.\n",
      "[bwa_aln_core] calculate SA coordinate... 7.53 sec\n",
      "[bwa_aln_core] write to the disk... 0.02 sec\n",
      "[bwa_aln_core] 1835008 sequences have been processed.\n",
      "[bwa_aln_core] calculate SA coordinate... 7.44 sec\n",
      "[bwa_aln_core] write to the disk... 0.02 sec\n",
      "[bwa_aln_core] 2097152 sequences have been processed.\n",
      "[bwa_aln_core] calculate SA coordinate... 7.52 sec\n",
      "[bwa_aln_core] write to the disk... 0.02 sec\n",
      "[bwa_aln_core] 2359296 sequences have been processed.\n",
      "[bwa_aln_core] calculate SA coordinate... 7.56 sec\n",
      "[bwa_aln_core] write to the disk... 0.02 sec\n",
      "[bwa_aln_core] 2621440 sequences have been processed.\n",
      "[bwa_aln_core] calculate SA coordinate... 1.17 sec\n",
      "[bwa_aln_core] write to the disk... 0.00 sec\n",
      "[bwa_aln_core] 2661285 sequences have been processed.\n",
      "[main] Version: 0.7.17-r1188\n",
      "[main] CMD: /home/will/src_and_bin/Other_Software/miniconda3/envs/tnseq_avium_abx_fordistro/bin/bwa aln input/Genome/GCF_000195955.2_ASM19595v2_genomic.fna output/tpp_out/SRR23906893.genomic2\n",
      "[main] Real time: 77.839 sec; CPU: 77.744 sec\n",
      "\n",
      "[tn_preprocess] /home/will/src_and_bin/Other_Software/miniconda3/envs/tnseq_avium_abx_fordistro/bin/bwa sampe input/Genome/GCF_000195955.2_ASM19595v2_genomic.fna output/tpp_out/SRR23906893.sai1 output/tpp_out/SRR23906893.sai2 output/tpp_out/SRR23906893.trimmed1 output/tpp_out/SRR23906893.genomic2 > output/tpp_out/SRR23906893.sam\n",
      "[tn_preprocess] 2000000 reads processed\n",
      "[bwa_sai2sam_pe_core] convert to sequence coordinate...\n",
      "[infer_isize] (25, 50, 75) percentile: (120, 186, 247)\n",
      "[infer_isize] low and high boundaries: 100 and 501 for estimating avg and std\n",
      "[infer_isize] inferred external isize from 189213 pairs: 214.548 +/- 74.009\n",
      "[infer_isize] skewness: 0.704; kurtosis: 0.335; ap_prior: 1.05e-05\n",
      "[infer_isize] inferred maximum insert size: 654 (5.94 sigma)\n",
      "[bwa_sai2sam_pe_core] time elapses: 1.50 sec\n",
      "[bwa_sai2sam_pe_core] changing coordinates of 867 alignments.\n",
      "[bwa_sai2sam_pe_core] align unmapped mate...\n",
      "[bwa_paired_sw] 2233 out of 19395 Q17 singletons are mated.\n",
      "[bwa_paired_sw] 42 out of 41422 Q17 discordant pairs are fixed.\n",
      "[bwa_sai2sam_pe_core] time elapses: 1.01 sec\n",
      "[bwa_sai2sam_pe_core] refine gapped alignments... 0.28 sec\n",
      "[bwa_sai2sam_pe_core] print alignments... 0.62 sec\n",
      "[bwa_sai2sam_pe_core] 262144 sequences have been processed.\n",
      "[bwa_sai2sam_pe_core] convert to sequence coordinate...\n",
      "[infer_isize] (25, 50, 75) percentile: (117, 182, 244)\n",
      "[infer_isize] low and high boundaries: 100 and 498 for estimating avg and std\n",
      "[infer_isize] inferred external isize from 185664 pairs: 212.533 +/- 73.205\n",
      "[infer_isize] skewness: 0.700; kurtosis: 0.299; ap_prior: 1.00e-05\n",
      "[infer_isize] inferred maximum insert size: 647 (5.94 sigma)\n",
      "[bwa_sai2sam_pe_core] time elapses: 1.63 sec\n",
      "[bwa_sai2sam_pe_core] changing coordinates of 878 alignments.\n",
      "[bwa_sai2sam_pe_core] align unmapped mate...\n",
      "[bwa_paired_sw] 2590 out of 20626 Q17 singletons are mated.\n",
      "[bwa_paired_sw] 41 out of 42955 Q17 discordant pairs are fixed.\n",
      "[bwa_sai2sam_pe_core] time elapses: 1.05 sec\n",
      "[bwa_sai2sam_pe_core] refine gapped alignments... 0.28 sec\n",
      "[bwa_sai2sam_pe_core] print alignments... 0.64 sec\n",
      "[bwa_sai2sam_pe_core] 524288 sequences have been processed.\n",
      "[bwa_sai2sam_pe_core] convert to sequence coordinate...\n",
      "[infer_isize] (25, 50, 75) percentile: (117, 183, 245)\n",
      "[infer_isize] low and high boundaries: 100 and 501 for estimating avg and std\n",
      "[infer_isize] inferred external isize from 186500 pairs: 213.027 +/- 73.314\n",
      "[infer_isize] skewness: 0.693; kurtosis: 0.298; ap_prior: 1.00e-05\n",
      "[infer_isize] inferred maximum insert size: 649 (5.94 sigma)\n",
      "[bwa_sai2sam_pe_core] time elapses: 1.62 sec\n",
      "[bwa_sai2sam_pe_core] changing coordinates of 910 alignments.\n",
      "[bwa_sai2sam_pe_core] align unmapped mate...\n",
      "[bwa_paired_sw] 2475 out of 20289 Q17 singletons are mated.\n",
      "[bwa_paired_sw] 30 out of 42614 Q17 discordant pairs are fixed.\n",
      "[bwa_sai2sam_pe_core] time elapses: 1.04 sec\n",
      "[bwa_sai2sam_pe_core] refine gapped alignments... 0.28 sec\n",
      "[bwa_sai2sam_pe_core] print alignments... 0.64 sec\n",
      "[bwa_sai2sam_pe_core] 786432 sequences have been processed.\n",
      "[bwa_sai2sam_pe_core] convert to sequence coordinate...\n",
      "[infer_isize] (25, 50, 75) percentile: (119, 185, 247)\n",
      "[infer_isize] low and high boundaries: 100 and 503 for estimating avg and std\n",
      "[infer_isize] inferred external isize from 188654 pairs: 214.251 +/- 74.026\n",
      "[infer_isize] skewness: 0.709; kurtosis: 0.344; ap_prior: 1.00e-05\n",
      "[infer_isize] inferred maximum insert size: 654 (5.94 sigma)\n",
      "[bwa_sai2sam_pe_core] time elapses: 1.62 sec\n",
      "[bwa_sai2sam_pe_core] changing coordinates of 924 alignments.\n",
      "[bwa_sai2sam_pe_core] align unmapped mate...\n",
      "[bwa_paired_sw] 2256 out of 19381 Q17 singletons are mated.\n",
      "[bwa_paired_sw] 36 out of 41934 Q17 discordant pairs are fixed.\n",
      "[bwa_sai2sam_pe_core] time elapses: 1.01 sec\n",
      "[bwa_sai2sam_pe_core] refine gapped alignments... 0.27 sec\n",
      "[bwa_sai2sam_pe_core] print alignments... 0.63 sec\n",
      "[bwa_sai2sam_pe_core] 1048576 sequences have been processed.\n",
      "[bwa_sai2sam_pe_core] convert to sequence coordinate...\n",
      "[infer_isize] (25, 50, 75) percentile: (115, 180, 243)\n",
      "[infer_isize] low and high boundaries: 100 and 499 for estimating avg and std\n",
      "[infer_isize] inferred external isize from 184112 pairs: 211.537 +/- 72.948\n",
      "[infer_isize] skewness: 0.716; kurtosis: 0.344; ap_prior: 1.00e-05\n",
      "[infer_isize] inferred maximum insert size: 645 (5.94 sigma)\n",
      "[bwa_sai2sam_pe_core] time elapses: 1.58 sec\n",
      "[bwa_sai2sam_pe_core] changing coordinates of 889 alignments.\n",
      "[bwa_sai2sam_pe_core] align unmapped mate...\n",
      "[bwa_paired_sw] 2720 out of 20880 Q17 singletons are mated.\n",
      "[bwa_paired_sw] 32 out of 44126 Q17 discordant pairs are fixed.\n",
      "[bwa_sai2sam_pe_core] time elapses: 1.07 sec\n",
      "[bwa_sai2sam_pe_core] refine gapped alignments... 0.28 sec\n",
      "[bwa_sai2sam_pe_core] print alignments... 0.63 sec\n",
      "[bwa_sai2sam_pe_core] 1310720 sequences have been processed.\n",
      "[bwa_sai2sam_pe_core] convert to sequence coordinate...\n",
      "[infer_isize] (25, 50, 75) percentile: (119, 186, 247)\n",
      "[infer_isize] low and high boundaries: 100 and 503 for estimating avg and std\n",
      "[infer_isize] inferred external isize from 189007 pairs: 214.412 +/- 74.027\n",
      "[infer_isize] skewness: 0.693; kurtosis: 0.315; ap_prior: 1.06e-05\n",
      "[infer_isize] inferred maximum insert size: 654 (5.94 sigma)\n",
      "[bwa_sai2sam_pe_core] time elapses: 1.62 sec\n",
      "[bwa_sai2sam_pe_core] changing coordinates of 851 alignments.\n",
      "[bwa_sai2sam_pe_core] align unmapped mate...\n",
      "[bwa_paired_sw] 2463 out of 19709 Q17 singletons are mated.\n",
      "[bwa_paired_sw] 24 out of 41423 Q17 discordant pairs are fixed.\n",
      "[bwa_sai2sam_pe_core] time elapses: 1.02 sec\n",
      "[bwa_sai2sam_pe_core] refine gapped alignments... 0.28 sec\n",
      "[bwa_sai2sam_pe_core] print alignments... 0.65 sec\n",
      "[bwa_sai2sam_pe_core] 1572864 sequences have been processed.\n",
      "[bwa_sai2sam_pe_core] convert to sequence coordinate...\n",
      "[infer_isize] (25, 50, 75) percentile: (117, 183, 245)\n",
      "[infer_isize] low and high boundaries: 100 and 501 for estimating avg and std\n",
      "[infer_isize] inferred external isize from 186594 pairs: 212.730 +/- 73.256\n",
      "[infer_isize] skewness: 0.695; kurtosis: 0.317; ap_prior: 1.00e-05\n",
      "[infer_isize] inferred maximum insert size: 648 (5.94 sigma)\n",
      "[bwa_sai2sam_pe_core] time elapses: 1.54 sec\n",
      "[bwa_sai2sam_pe_core] changing coordinates of 904 alignments.\n",
      "[bwa_sai2sam_pe_core] align unmapped mate...\n",
      "[bwa_paired_sw] 2695 out of 20496 Q17 singletons are mated.\n",
      "[bwa_paired_sw] 31 out of 42690 Q17 discordant pairs are fixed.\n",
      "[bwa_sai2sam_pe_core] time elapses: 1.04 sec\n",
      "[bwa_sai2sam_pe_core] refine gapped alignments... 0.28 sec\n",
      "[bwa_sai2sam_pe_core] print alignments... 0.67 sec\n",
      "[bwa_sai2sam_pe_core] 1835008 sequences have been processed.\n",
      "[bwa_sai2sam_pe_core] convert to sequence coordinate...\n",
      "[infer_isize] (25, 50, 75) percentile: (117, 183, 245)\n",
      "[infer_isize] low and high boundaries: 100 and 501 for estimating avg and std\n",
      "[infer_isize] inferred external isize from 185526 pairs: 212.950 +/- 73.413\n",
      "[infer_isize] skewness: 0.711; kurtosis: 0.362; ap_prior: 1.00e-05\n",
      "[infer_isize] inferred maximum insert size: 649 (5.94 sigma)\n",
      "[bwa_sai2sam_pe_core] time elapses: 1.61 sec\n",
      "[bwa_sai2sam_pe_core] changing coordinates of 827 alignments.\n",
      "[bwa_sai2sam_pe_core] align unmapped mate...\n",
      "[bwa_paired_sw] 2902 out of 20779 Q17 singletons are mated.\n",
      "[bwa_paired_sw] 28 out of 42757 Q17 discordant pairs are fixed.\n",
      "[bwa_sai2sam_pe_core] time elapses: 1.06 sec\n",
      "[bwa_sai2sam_pe_core] refine gapped alignments... 0.27 sec\n",
      "[bwa_sai2sam_pe_core] print alignments... 0.62 sec\n",
      "[bwa_sai2sam_pe_core] 2097152 sequences have been processed.\n",
      "[bwa_sai2sam_pe_core] convert to sequence coordinate...\n",
      "[infer_isize] (25, 50, 75) percentile: (118, 185, 247)\n",
      "[infer_isize] low and high boundaries: 100 and 505 for estimating avg and std\n",
      "[infer_isize] inferred external isize from 187905 pairs: 214.476 +/- 74.182\n",
      "[infer_isize] skewness: 0.705; kurtosis: 0.339; ap_prior: 1.00e-05\n",
      "[infer_isize] inferred maximum insert size: 655 (5.94 sigma)\n",
      "[bwa_sai2sam_pe_core] time elapses: 1.56 sec\n",
      "[bwa_sai2sam_pe_core] changing coordinates of 943 alignments.\n",
      "[bwa_sai2sam_pe_core] align unmapped mate...\n",
      "[bwa_paired_sw] 2552 out of 20156 Q17 singletons are mated.\n",
      "[bwa_paired_sw] 33 out of 41993 Q17 discordant pairs are fixed.\n",
      "[bwa_sai2sam_pe_core] time elapses: 1.05 sec\n",
      "[bwa_sai2sam_pe_core] refine gapped alignments... 0.30 sec\n",
      "[bwa_sai2sam_pe_core] print alignments... 0.66 sec\n",
      "[bwa_sai2sam_pe_core] 2359296 sequences have been processed.\n",
      "[bwa_sai2sam_pe_core] convert to sequence coordinate...\n",
      "[infer_isize] (25, 50, 75) percentile: (115, 181, 243)\n",
      "[infer_isize] low and high boundaries: 100 and 499 for estimating avg and std\n",
      "[infer_isize] inferred external isize from 183888 pairs: 211.531 +/- 72.759\n",
      "[infer_isize] skewness: 0.701; kurtosis: 0.310; ap_prior: 1.00e-05\n",
      "[infer_isize] inferred maximum insert size: 644 (5.94 sigma)\n",
      "[bwa_sai2sam_pe_core] time elapses: 1.83 sec\n",
      "[bwa_sai2sam_pe_core] changing coordinates of 825 alignments.\n",
      "[bwa_sai2sam_pe_core] align unmapped mate...\n",
      "[bwa_paired_sw] 3101 out of 21505 Q17 singletons are mated.\n",
      "[bwa_paired_sw] 36 out of 43481 Q17 discordant pairs are fixed.\n",
      "[bwa_sai2sam_pe_core] time elapses: 1.19 sec\n",
      "[bwa_sai2sam_pe_core] refine gapped alignments... 0.28 sec\n",
      "[bwa_sai2sam_pe_core] print alignments... 0.66 sec\n",
      "[bwa_sai2sam_pe_core] 2621440 sequences have been processed.\n",
      "[bwa_sai2sam_pe_core] convert to sequence coordinate...\n",
      "[infer_isize] (25, 50, 75) percentile: (115, 181, 243)\n",
      "[infer_isize] low and high boundaries: 100 and 499 for estimating avg and std\n",
      "[infer_isize] inferred external isize from 27922 pairs: 211.758 +/- 73.091\n",
      "[infer_isize] skewness: 0.707; kurtosis: 0.347; ap_prior: 1.00e-05\n",
      "[infer_isize] inferred maximum insert size: 646 (5.94 sigma)\n",
      "[bwa_sai2sam_pe_core] time elapses: 0.23 sec\n",
      "[bwa_sai2sam_pe_core] changing coordinates of 112 alignments.\n",
      "[bwa_sai2sam_pe_core] align unmapped mate...\n",
      "[bwa_paired_sw] 474 out of 3283 Q17 singletons are mated.\n",
      "[bwa_paired_sw] 4 out of 6546 Q17 discordant pairs are fixed.\n",
      "[bwa_sai2sam_pe_core] time elapses: 0.16 sec\n",
      "[bwa_sai2sam_pe_core] refine gapped alignments... 0.04 sec\n",
      "[bwa_sai2sam_pe_core] print alignments... 0.10 sec\n",
      "[bwa_sai2sam_pe_core] 2661285 sequences have been processed.\n",
      "[main] Version: 0.7.17-r1188\n",
      "[main] CMD: /home/will/src_and_bin/Other_Software/miniconda3/envs/tnseq_avium_abx_fordistro/bin/bwa sampe input/Genome/GCF_000195955.2_ASM19595v2_genomic.fna output/tpp_out/SRR23906893.sai1 output/tpp_out/SRR23906893.sai2 output/tpp_out/SRR23906893.trimmed1 output/tpp_out/SRR23906893.genomic2\n",
      "[main] Real time: 39.987 sec; CPU: 39.690 sec\n",
      "\n",
      "[tn_preprocess] tabulating template counts and statistics for reference genome input/Genome/GCF_000195955.2_ASM19595v2_genomic.fna\n",
      "[tn_preprocess] writing output/tpp_out/SRR23906893.wig\n",
      "[tn_preprocess] writing output/tpp_out/SRR23906893.tn_stats\n",
      "[tn_preprocess] Done.\n",
      "[tn_preprocess] 3000000 reads processed\n",
      "[tn_preprocess] 4000000 reads processed\n",
      "[tn_preprocess] creating output/tpp_out/SRR23906897.barcodes1\n",
      "[tn_preprocess] mapping reads using BWA...(this takes a couple of minutes)\n",
      "[tn_preprocess] /home/will/src_and_bin/Other_Software/miniconda3/envs/tnseq_avium_abx_fordistro/bin/bwa aln input/Genome/GCF_000195955.2_ASM19595v2_genomic.fna output/tpp_out/SRR23906897.trimmed1 > output/tpp_out/SRR23906897.sai1\n",
      "[bwa_aln] 17bp reads: max_diff = 2\n",
      "[bwa_aln] 38bp reads: max_diff = 3\n",
      "[bwa_aln] 64bp reads: max_diff = 4\n",
      "[bwa_aln] 93bp reads: max_diff = 5\n",
      "[bwa_aln] 124bp reads: max_diff = 6\n",
      "[bwa_aln] 157bp reads: max_diff = 7\n",
      "[bwa_aln] 190bp reads: max_diff = 8\n",
      "[bwa_aln] 225bp reads: max_diff = 9\n",
      "[bwa_aln_core] calculate SA coordinate... 6.98 sec\n",
      "[bwa_aln_core] write to the disk... 0.02 sec\n",
      "[bwa_aln_core] 262144 sequences have been processed.\n",
      "[bwa_aln_core] calculate SA coordinate... 7.19 sec\n",
      "[bwa_aln_core] write to the disk... 0.02 sec\n",
      "[bwa_aln_core] 524288 sequences have been processed.\n",
      "[bwa_aln_core] calculate SA coordinate... 7.15 sec\n",
      "[bwa_aln_core] write to the disk... 0.02 sec\n",
      "[bwa_aln_core] 786432 sequences have been processed.\n",
      "[bwa_aln_core] calculate SA coordinate... 7.44 sec\n",
      "[bwa_aln_core] write to the disk... 0.02 sec\n",
      "[bwa_aln_core] 1048576 sequences have been processed.\n",
      "[bwa_aln_core] calculate SA coordinate... 7.00 sec\n",
      "[bwa_aln_core] write to the disk... 0.02 sec\n",
      "[bwa_aln_core] 1310720 sequences have been processed.\n",
      "[bwa_aln_core] calculate SA coordinate... 7.41 sec\n",
      "[bwa_aln_core] write to the disk... 0.02 sec\n",
      "[bwa_aln_core] 1572864 sequences have been processed.\n",
      "[bwa_aln_core] calculate SA coordinate... 7.72 sec\n",
      "[bwa_aln_core] write to the disk... 0.02 sec\n",
      "[bwa_aln_core] 1835008 sequences have been processed.\n",
      "[bwa_aln_core] calculate SA coordinate... 7.49 sec\n",
      "[bwa_aln_core] write to the disk... 0.02 sec\n",
      "[bwa_aln_core] 2097152 sequences have been processed.\n",
      "[bwa_aln_core] calculate SA coordinate... 7.63 sec\n",
      "[bwa_aln_core] write to the disk... 0.02 sec\n",
      "[bwa_aln_core] 2359296 sequences have been processed.\n",
      "[bwa_aln_core] calculate SA coordinate... 7.21 sec\n",
      "[bwa_aln_core] write to the disk... 0.02 sec\n",
      "[bwa_aln_core] 2621440 sequences have been processed.\n",
      "[bwa_aln_core] calculate SA coordinate... 7.01 sec\n",
      "[bwa_aln_core] write to the disk... 0.02 sec\n",
      "[bwa_aln_core] 2883584 sequences have been processed.\n",
      "[bwa_aln_core] calculate SA coordinate... 7.03 sec\n",
      "[bwa_aln_core] write to the disk... 0.02 sec\n",
      "[bwa_aln_core] 3145728 sequences have been processed.\n",
      "[bwa_aln_core] calculate SA coordinate... 8.15 sec\n",
      "[bwa_aln_core] write to the disk... 0.02 sec\n",
      "[bwa_aln_core] 3407872 sequences have been processed.\n",
      "[bwa_aln_core] calculate SA coordinate... 6.12 sec\n",
      "[bwa_aln_core] write to the disk... 0.02 sec\n",
      "[bwa_aln_core] 3670016 sequences have been processed.\n",
      "[bwa_aln_core] calculate SA coordinate... 7.43 sec\n",
      "[bwa_aln_core] write to the disk... 0.02 sec\n",
      "[bwa_aln_core] 3932160 sequences have been processed.\n",
      "[bwa_aln_core] calculate SA coordinate... 7.61 sec\n",
      "[bwa_aln_core] write to the disk... 0.02 sec\n",
      "[bwa_aln_core] 4194304 sequences have been processed.\n",
      "[bwa_aln_core] calculate SA coordinate... 8.38 sec\n",
      "[bwa_aln_core] write to the disk... 0.02 sec\n",
      "[bwa_aln_core] 4456448 sequences have been processed.\n",
      "[bwa_aln_core] calculate SA coordinate... 8.43 sec\n",
      "[bwa_aln_core] write to the disk... 0.02 sec\n",
      "[bwa_aln_core] 4718592 sequences have been processed.\n",
      "[bwa_aln_core] calculate SA coordinate... 5.18 sec\n",
      "[bwa_aln_core] write to the disk... 0.01 sec\n",
      "[bwa_aln_core] 4886129 sequences have been processed.\n",
      "[main] Version: 0.7.17-r1188\n",
      "[main] CMD: /home/will/src_and_bin/Other_Software/miniconda3/envs/tnseq_avium_abx_fordistro/bin/bwa aln input/Genome/GCF_000195955.2_ASM19595v2_genomic.fna output/tpp_out/SRR23906897.trimmed1\n",
      "[main] Real time: 141.617 sec; CPU: 141.481 sec\n",
      "\n",
      "[tn_preprocess] /home/will/src_and_bin/Other_Software/miniconda3/envs/tnseq_avium_abx_fordistro/bin/bwa aln input/Genome/GCF_000195955.2_ASM19595v2_genomic.fna output/tpp_out/SRR23906897.genomic2 > output/tpp_out/SRR23906897.sai2\n",
      "[bwa_aln] 17bp reads: max_diff = 2\n",
      "[bwa_aln] 38bp reads: max_diff = 3\n",
      "[bwa_aln] 64bp reads: max_diff = 4\n",
      "[bwa_aln] 93bp reads: max_diff = 5\n",
      "[bwa_aln] 124bp reads: max_diff = 6\n",
      "[bwa_aln] 157bp reads: max_diff = 7\n",
      "[bwa_aln] 190bp reads: max_diff = 8\n",
      "[bwa_aln] 225bp reads: max_diff = 9\n",
      "[bwa_aln_core] calculate SA coordinate... 6.46 sec\n",
      "[bwa_aln_core] write to the disk... 0.02 sec\n",
      "[bwa_aln_core] 262144 sequences have been processed.\n",
      "[bwa_aln_core] calculate SA coordinate... 7.16 sec\n",
      "[bwa_aln_core] write to the disk... 0.02 sec\n",
      "[bwa_aln_core] 524288 sequences have been processed.\n",
      "[bwa_aln_core] calculate SA coordinate... 7.38 sec\n",
      "[bwa_aln_core] write to the disk... 0.02 sec\n",
      "[bwa_aln_core] 786432 sequences have been processed.\n",
      "[bwa_aln_core] calculate SA coordinate... 5.12 sec\n",
      "[bwa_aln_core] write to the disk... 0.01 sec\n",
      "[bwa_aln_core] 1048576 sequences have been processed.\n",
      "[bwa_aln_core] calculate SA coordinate... 5.84 sec\n",
      "[bwa_aln_core] write to the disk... 0.02 sec\n",
      "[bwa_aln_core] 1310720 sequences have been processed.\n",
      "[bwa_aln_core] calculate SA coordinate... 7.02 sec\n",
      "[bwa_aln_core] write to the disk... 0.02 sec\n",
      "[bwa_aln_core] 1572864 sequences have been processed.\n",
      "[bwa_aln_core] calculate SA coordinate... 7.23 sec\n",
      "[bwa_aln_core] write to the disk... 0.02 sec\n",
      "[bwa_aln_core] 1835008 sequences have been processed.\n",
      "[bwa_aln_core] calculate SA coordinate... 8.69 sec\n",
      "[bwa_aln_core] write to the disk... 0.02 sec\n",
      "[bwa_aln_core] 2097152 sequences have been processed.\n",
      "[bwa_aln_core] calculate SA coordinate... 5.30 sec\n",
      "[bwa_aln_core] write to the disk... 0.02 sec\n",
      "[bwa_aln_core] 2359296 sequences have been processed.\n",
      "[bwa_aln_core] calculate SA coordinate... 6.68 sec\n",
      "[bwa_aln_core] write to the disk... 0.02 sec\n",
      "[bwa_aln_core] 2621440 sequences have been processed.\n",
      "[bwa_aln_core] calculate SA coordinate... 6.43 sec\n",
      "[bwa_aln_core] write to the disk... 0.02 sec\n",
      "[bwa_aln_core] 2883584 sequences have been processed.\n",
      "[bwa_aln_core] calculate SA coordinate... 6.49 sec\n",
      "[bwa_aln_core] write to the disk... 0.02 sec\n",
      "[bwa_aln_core] 3145728 sequences have been processed.\n",
      "[bwa_aln_core] calculate SA coordinate... 10.66 sec\n",
      "[bwa_aln_core] write to the disk... 0.02 sec\n",
      "[bwa_aln_core] 3407872 sequences have been processed.\n",
      "[bwa_aln_core] calculate SA coordinate... 5.78 sec\n",
      "[bwa_aln_core] write to the disk... 0.02 sec\n",
      "[bwa_aln_core] 3670016 sequences have been processed.\n",
      "[bwa_aln_core] calculate SA coordinate... 6.19 sec\n",
      "[bwa_aln_core] write to the disk... 0.02 sec\n",
      "[bwa_aln_core] 3932160 sequences have been processed.\n",
      "[bwa_aln_core] calculate SA coordinate... 6.52 sec\n",
      "[bwa_aln_core] write to the disk... 0.02 sec\n",
      "[bwa_aln_core] 4194304 sequences have been processed.\n",
      "[bwa_aln_core] calculate SA coordinate... 6.40 sec\n",
      "[bwa_aln_core] write to the disk... 0.02 sec\n",
      "[bwa_aln_core] 4456448 sequences have been processed.\n",
      "[bwa_aln_core] calculate SA coordinate... 5.34 sec\n",
      "[bwa_aln_core] write to the disk... 0.02 sec\n",
      "[bwa_aln_core] 4718592 sequences have been processed.\n",
      "[bwa_aln_core] calculate SA coordinate... 2.33 sec\n",
      "[bwa_aln_core] write to the disk... 0.01 sec\n",
      "[bwa_aln_core] 4886129 sequences have been processed.\n",
      "[main] Version: 0.7.17-r1188\n",
      "[main] CMD: /home/will/src_and_bin/Other_Software/miniconda3/envs/tnseq_avium_abx_fordistro/bin/bwa aln input/Genome/GCF_000195955.2_ASM19595v2_genomic.fna output/tpp_out/SRR23906897.genomic2\n",
      "[main] Real time: 125.821 sec; CPU: 125.654 sec\n",
      "\n",
      "[tn_preprocess] /home/will/src_and_bin/Other_Software/miniconda3/envs/tnseq_avium_abx_fordistro/bin/bwa sampe input/Genome/GCF_000195955.2_ASM19595v2_genomic.fna output/tpp_out/SRR23906897.sai1 output/tpp_out/SRR23906897.sai2 output/tpp_out/SRR23906897.trimmed1 output/tpp_out/SRR23906897.genomic2 > output/tpp_out/SRR23906897.sam\n",
      "[bwa_sai2sam_pe_core] convert to sequence coordinate...\n",
      "[infer_isize] (25, 50, 75) percentile: (130, 194, 253)\n",
      "[infer_isize] low and high boundaries: 100 and 499 for estimating avg and std\n",
      "[infer_isize] inferred external isize from 185885 pairs: 217.978 +/- 74.296\n",
      "[infer_isize] skewness: 0.686; kurtosis: 0.314; ap_prior: 1.00e-05\n",
      "[infer_isize] inferred maximum insert size: 659 (5.94 sigma)\n",
      "[bwa_sai2sam_pe_core] time elapses: 1.35 sec\n",
      "[bwa_sai2sam_pe_core] changing coordinates of 946 alignments.\n",
      "[bwa_sai2sam_pe_core] align unmapped mate...\n",
      "[bwa_paired_sw] 5323 out of 24885 Q17 singletons are mated.\n",
      "[bwa_paired_sw] 33 out of 32858 Q17 discordant pairs are fixed.\n",
      "[bwa_sai2sam_pe_core] time elapses: 0.91 sec\n",
      "[bwa_sai2sam_pe_core] refine gapped alignments... 0.25 sec\n",
      "[bwa_sai2sam_pe_core] print alignments... 0.60 sec\n",
      "[bwa_sai2sam_pe_core] 262144 sequences have been processed.\n",
      "[bwa_sai2sam_pe_core] convert to sequence coordinate...\n",
      "[infer_isize] (25, 50, 75) percentile: (129, 193, 251)\n",
      "[infer_isize] low and high boundaries: 100 and 495 for estimating avg and std\n",
      "[infer_isize] inferred external isize from 178855 pairs: 216.775 +/- 73.561\n",
      "[infer_isize] skewness: 0.687; kurtosis: 0.320; ap_prior: 1.00e-05\n",
      "[infer_isize] inferred maximum insert size: 654 (5.94 sigma)\n",
      "[bwa_sai2sam_pe_core] time elapses: 1.33 sec\n",
      "[bwa_sai2sam_pe_core] changing coordinates of 879 alignments.\n",
      "[bwa_sai2sam_pe_core] align unmapped mate...\n",
      "[bwa_paired_sw] 4949 out of 31188 Q17 singletons are mated.\n",
      "[bwa_paired_sw] 20 out of 32110 Q17 discordant pairs are fixed.\n",
      "[bwa_sai2sam_pe_core] time elapses: 0.88 sec\n",
      "[bwa_sai2sam_pe_core] refine gapped alignments... 0.25 sec\n",
      "[bwa_sai2sam_pe_core] print alignments... 0.59 sec\n",
      "[bwa_sai2sam_pe_core] 524288 sequences have been processed.\n",
      "[bwa_sai2sam_pe_core] convert to sequence coordinate...\n",
      "[infer_isize] (25, 50, 75) percentile: (128, 192, 250)\n",
      "[infer_isize] low and high boundaries: 100 and 494 for estimating avg and std\n",
      "[infer_isize] inferred external isize from 176042 pairs: 216.044 +/- 73.062\n",
      "[infer_isize] skewness: 0.688; kurtosis: 0.324; ap_prior: 1.00e-05\n",
      "[infer_isize] inferred maximum insert size: 650 (5.94 sigma)\n",
      "[bwa_sai2sam_pe_core] time elapses: 1.54 sec\n",
      "[bwa_sai2sam_pe_core] changing coordinates of 825 alignments.\n",
      "[bwa_sai2sam_pe_core] align unmapped mate...\n",
      "[bwa_paired_sw] 7053 out of 31227 Q17 singletons are mated.\n",
      "[bwa_paired_sw] 27 out of 32224 Q17 discordant pairs are fixed.\n",
      "[bwa_sai2sam_pe_core] time elapses: 0.96 sec\n",
      "[bwa_sai2sam_pe_core] refine gapped alignments... 0.25 sec\n",
      "[bwa_sai2sam_pe_core] print alignments... 0.59 sec\n",
      "[bwa_sai2sam_pe_core] 786432 sequences have been processed.\n",
      "[bwa_sai2sam_pe_core] convert to sequence coordinate...\n",
      "[infer_isize] (25, 50, 75) percentile: (127, 189, 247)\n",
      "[infer_isize] low and high boundaries: 100 and 487 for estimating avg and std\n",
      "[infer_isize] inferred external isize from 123215 pairs: 213.493 +/- 72.241\n",
      "[infer_isize] skewness: 0.692; kurtosis: 0.315; ap_prior: 1.00e-05\n",
      "[infer_isize] inferred maximum insert size: 643 (5.94 sigma)\n",
      "[bwa_sai2sam_pe_core] time elapses: 1.12 sec\n",
      "[bwa_sai2sam_pe_core] changing coordinates of 604 alignments.\n",
      "[bwa_sai2sam_pe_core] align unmapped mate...\n",
      "[bwa_paired_sw] 35533 out of 82454 Q17 singletons are mated.\n",
      "[bwa_paired_sw] 19 out of 22604 Q17 discordant pairs are fixed.\n",
      "[bwa_sai2sam_pe_core] time elapses: 1.74 sec\n",
      "[bwa_sai2sam_pe_core] refine gapped alignments... 0.24 sec\n",
      "[bwa_sai2sam_pe_core] print alignments... 0.56 sec\n",
      "[bwa_sai2sam_pe_core] 1048576 sequences have been processed.\n",
      "[bwa_sai2sam_pe_core] convert to sequence coordinate...\n",
      "[infer_isize] (25, 50, 75) percentile: (129, 192, 252)\n",
      "[infer_isize] low and high boundaries: 100 and 498 for estimating avg and std\n",
      "[infer_isize] inferred external isize from 153068 pairs: 216.968 +/- 74.249\n",
      "[infer_isize] skewness: 0.706; kurtosis: 0.359; ap_prior: 1.00e-05\n",
      "[infer_isize] inferred maximum insert size: 658 (5.94 sigma)\n",
      "[bwa_sai2sam_pe_core] time elapses: 1.30 sec\n",
      "[bwa_sai2sam_pe_core] changing coordinates of 703 alignments.\n",
      "[bwa_sai2sam_pe_core] align unmapped mate...\n",
      "[bwa_paired_sw] 18927 out of 53876 Q17 singletons are mated.\n",
      "[bwa_paired_sw] 17 out of 27228 Q17 discordant pairs are fixed.\n",
      "[bwa_sai2sam_pe_core] time elapses: 1.24 sec\n",
      "[bwa_sai2sam_pe_core] refine gapped alignments... 0.24 sec\n",
      "[bwa_sai2sam_pe_core] print alignments... 0.58 sec\n",
      "[bwa_sai2sam_pe_core] 1310720 sequences have been processed.\n",
      "[bwa_sai2sam_pe_core] convert to sequence coordinate...\n",
      "[infer_isize] (25, 50, 75) percentile: (130, 194, 253)\n",
      "[infer_isize] low and high boundaries: 100 and 499 for estimating avg and std\n",
      "[infer_isize] inferred external isize from 183809 pairs: 218.058 +/- 74.406\n",
      "[infer_isize] skewness: 0.695; kurtosis: 0.332; ap_prior: 1.00e-05\n",
      "[infer_isize] inferred maximum insert size: 660 (5.94 sigma)\n",
      "[bwa_sai2sam_pe_core] time elapses: 1.42 sec\n",
      "[bwa_sai2sam_pe_core] changing coordinates of 941 alignments.\n",
      "[bwa_sai2sam_pe_core] align unmapped mate...\n",
      "[bwa_paired_sw] 4870 out of 26716 Q17 singletons are mated.\n",
      "[bwa_paired_sw] 17 out of 32390 Q17 discordant pairs are fixed.\n",
      "[bwa_sai2sam_pe_core] time elapses: 0.89 sec\n",
      "[bwa_sai2sam_pe_core] refine gapped alignments... 0.25 sec\n",
      "[bwa_sai2sam_pe_core] print alignments... 0.60 sec\n",
      "[bwa_sai2sam_pe_core] 1572864 sequences have been processed.\n",
      "[bwa_sai2sam_pe_core] convert to sequence coordinate...\n",
      "[infer_isize] (25, 50, 75) percentile: (130, 193, 252)\n",
      "[infer_isize] low and high boundaries: 100 and 496 for estimating avg and std\n",
      "[infer_isize] inferred external isize from 182883 pairs: 217.306 +/- 73.763\n",
      "[infer_isize] skewness: 0.680; kurtosis: 0.294; ap_prior: 1.00e-05\n",
      "[infer_isize] inferred maximum insert size: 655 (5.94 sigma)\n",
      "[bwa_sai2sam_pe_core] time elapses: 1.43 sec\n",
      "[bwa_sai2sam_pe_core] changing coordinates of 868 alignments.\n",
      "[bwa_sai2sam_pe_core] align unmapped mate...\n",
      "[bwa_paired_sw] 5299 out of 27512 Q17 singletons are mated.\n",
      "[bwa_paired_sw] 20 out of 32280 Q17 discordant pairs are fixed.\n",
      "[bwa_sai2sam_pe_core] time elapses: 0.90 sec\n",
      "[bwa_sai2sam_pe_core] refine gapped alignments... 0.25 sec\n",
      "[bwa_sai2sam_pe_core] print alignments... 0.61 sec\n",
      "[bwa_sai2sam_pe_core] 1835008 sequences have been processed.\n",
      "[bwa_sai2sam_pe_core] convert to sequence coordinate...\n",
      "[infer_isize] (25, 50, 75) percentile: (129, 192, 250)\n",
      "[infer_isize] low and high boundaries: 100 and 492 for estimating avg and std\n",
      "[infer_isize] inferred external isize from 165157 pairs: 215.669 +/- 73.045\n",
      "[infer_isize] skewness: 0.691; kurtosis: 0.318; ap_prior: 1.00e-05\n",
      "[infer_isize] inferred maximum insert size: 650 (5.94 sigma)\n",
      "[bwa_sai2sam_pe_core] time elapses: 1.27 sec\n",
      "[bwa_sai2sam_pe_core] changing coordinates of 763 alignments.\n",
      "[bwa_sai2sam_pe_core] align unmapped mate...\n",
      "[bwa_paired_sw] 13575 out of 39802 Q17 singletons are mated.\n",
      "[bwa_paired_sw] 22 out of 29369 Q17 discordant pairs are fixed.\n",
      "[bwa_sai2sam_pe_core] time elapses: 1.16 sec\n",
      "[bwa_sai2sam_pe_core] refine gapped alignments... 0.25 sec\n",
      "[bwa_sai2sam_pe_core] print alignments... 0.58 sec\n",
      "[bwa_sai2sam_pe_core] 2097152 sequences have been processed.\n",
      "[bwa_sai2sam_pe_core] convert to sequence coordinate...\n",
      "[infer_isize] (25, 50, 75) percentile: (128, 190, 248)\n",
      "[infer_isize] low and high boundaries: 100 and 488 for estimating avg and std\n",
      "[infer_isize] inferred external isize from 140601 pairs: 214.227 +/- 72.247\n",
      "[infer_isize] skewness: 0.684; kurtosis: 0.294; ap_prior: 1.00e-05\n",
      "[infer_isize] inferred maximum insert size: 643 (5.94 sigma)\n",
      "[bwa_sai2sam_pe_core] time elapses: 1.24 sec\n",
      "[bwa_sai2sam_pe_core] changing coordinates of 715 alignments.\n",
      "[bwa_sai2sam_pe_core] align unmapped mate...\n",
      "[bwa_paired_sw] 22061 out of 62395 Q17 singletons are mated.\n",
      "[bwa_paired_sw] 17 out of 25074 Q17 discordant pairs are fixed.\n",
      "[bwa_sai2sam_pe_core] time elapses: 1.36 sec\n",
      "[bwa_sai2sam_pe_core] refine gapped alignments... 0.24 sec\n",
      "[bwa_sai2sam_pe_core] print alignments... 0.56 sec\n",
      "[bwa_sai2sam_pe_core] 2359296 sequences have been processed.\n",
      "[bwa_sai2sam_pe_core] convert to sequence coordinate...\n",
      "[infer_isize] (25, 50, 75) percentile: (130, 194, 253)\n",
      "[infer_isize] low and high boundaries: 100 and 499 for estimating avg and std\n",
      "[infer_isize] inferred external isize from 186244 pairs: 218.080 +/- 74.519\n",
      "[infer_isize] skewness: 0.690; kurtosis: 0.320; ap_prior: 1.00e-05\n",
      "[infer_isize] inferred maximum insert size: 661 (5.94 sigma)\n",
      "[bwa_sai2sam_pe_core] time elapses: 1.38 sec\n",
      "[bwa_sai2sam_pe_core] changing coordinates of 982 alignments.\n",
      "[bwa_sai2sam_pe_core] align unmapped mate...\n",
      "[bwa_paired_sw] 5099 out of 25686 Q17 singletons are mated.\n",
      "[bwa_paired_sw] 23 out of 32557 Q17 discordant pairs are fixed.\n",
      "[bwa_sai2sam_pe_core] time elapses: 0.90 sec\n",
      "[bwa_sai2sam_pe_core] refine gapped alignments... 0.25 sec\n",
      "[bwa_sai2sam_pe_core] print alignments... 0.61 sec\n",
      "[bwa_sai2sam_pe_core] 2621440 sequences have been processed.\n",
      "[bwa_sai2sam_pe_core] convert to sequence coordinate...\n",
      "[infer_isize] (25, 50, 75) percentile: (129, 193, 252)\n",
      "[infer_isize] low and high boundaries: 100 and 498 for estimating avg and std\n",
      "[infer_isize] inferred external isize from 184144 pairs: 217.263 +/- 73.981\n",
      "[infer_isize] skewness: 0.685; kurtosis: 0.317; ap_prior: 1.00e-05\n",
      "[infer_isize] inferred maximum insert size: 657 (5.94 sigma)\n",
      "[bwa_sai2sam_pe_core] time elapses: 1.41 sec\n",
      "[bwa_sai2sam_pe_core] changing coordinates of 940 alignments.\n",
      "[bwa_sai2sam_pe_core] align unmapped mate...\n",
      "[bwa_paired_sw] 5361 out of 26120 Q17 singletons are mated.\n",
      "[bwa_paired_sw] 20 out of 32746 Q17 discordant pairs are fixed.\n",
      "[bwa_sai2sam_pe_core] time elapses: 0.91 sec\n",
      "[bwa_sai2sam_pe_core] refine gapped alignments... 0.25 sec\n",
      "[bwa_sai2sam_pe_core] print alignments... 0.60 sec\n",
      "[bwa_sai2sam_pe_core] 2883584 sequences have been processed.\n",
      "[bwa_sai2sam_pe_core] convert to sequence coordinate...\n",
      "[infer_isize] (25, 50, 75) percentile: (129, 193, 252)\n",
      "[infer_isize] low and high boundaries: 100 and 498 for estimating avg and std\n",
      "[infer_isize] inferred external isize from 182948 pairs: 217.623 +/- 74.146\n",
      "[infer_isize] skewness: 0.689; kurtosis: 0.320; ap_prior: 1.00e-05\n",
      "[infer_isize] inferred maximum insert size: 658 (5.94 sigma)\n",
      "[bwa_sai2sam_pe_core] time elapses: 1.42 sec\n",
      "[bwa_sai2sam_pe_core] changing coordinates of 943 alignments.\n",
      "[bwa_sai2sam_pe_core] align unmapped mate...\n",
      "[bwa_paired_sw] 5924 out of 26739 Q17 singletons are mated.\n",
      "[bwa_paired_sw] 25 out of 32906 Q17 discordant pairs are fixed.\n",
      "[bwa_sai2sam_pe_core] time elapses: 0.94 sec\n",
      "[bwa_sai2sam_pe_core] refine gapped alignments... 0.25 sec\n",
      "[bwa_sai2sam_pe_core] print alignments... 0.60 sec\n",
      "[bwa_sai2sam_pe_core] 3145728 sequences have been processed.\n",
      "[bwa_sai2sam_pe_core] convert to sequence coordinate...\n",
      "[infer_isize] (25, 50, 75) percentile: (127, 190, 248)\n",
      "[infer_isize] low and high boundaries: 100 and 490 for estimating avg and std\n",
      "[infer_isize] inferred external isize from 123097 pairs: 214.119 +/- 72.363\n",
      "[infer_isize] skewness: 0.694; kurtosis: 0.337; ap_prior: 1.00e-05\n",
      "[infer_isize] inferred maximum insert size: 644 (5.94 sigma)\n",
      "[bwa_sai2sam_pe_core] time elapses: 1.14 sec\n",
      "[bwa_sai2sam_pe_core] changing coordinates of 605 alignments.\n",
      "[bwa_sai2sam_pe_core] align unmapped mate...\n",
      "[bwa_paired_sw] 41690 out of 85448 Q17 singletons are mated.\n",
      "[bwa_paired_sw] 18 out of 22858 Q17 discordant pairs are fixed.\n",
      "[bwa_sai2sam_pe_core] time elapses: 1.81 sec\n",
      "[bwa_sai2sam_pe_core] refine gapped alignments... 0.24 sec\n",
      "[bwa_sai2sam_pe_core] print alignments... 0.57 sec\n",
      "[bwa_sai2sam_pe_core] 3407872 sequences have been processed.\n",
      "[bwa_sai2sam_pe_core] convert to sequence coordinate...\n",
      "[infer_isize] (25, 50, 75) percentile: (130, 192, 250)\n",
      "[infer_isize] low and high boundaries: 100 and 490 for estimating avg and std\n",
      "[infer_isize] inferred external isize from 144331 pairs: 215.369 +/- 72.719\n",
      "[infer_isize] skewness: 0.675; kurtosis: 0.280; ap_prior: 1.00e-05\n",
      "[infer_isize] inferred maximum insert size: 647 (5.94 sigma)\n",
      "[bwa_sai2sam_pe_core] time elapses: 1.22 sec\n",
      "[bwa_sai2sam_pe_core] changing coordinates of 716 alignments.\n",
      "[bwa_sai2sam_pe_core] align unmapped mate...\n",
      "[bwa_paired_sw] 19916 out of 48080 Q17 singletons are mated.\n",
      "[bwa_paired_sw] 17 out of 24862 Q17 discordant pairs are fixed.\n",
      "[bwa_sai2sam_pe_core] time elapses: 1.26 sec\n",
      "[bwa_sai2sam_pe_core] refine gapped alignments... 0.23 sec\n",
      "[bwa_sai2sam_pe_core] print alignments... 0.55 sec\n",
      "[bwa_sai2sam_pe_core] 3670016 sequences have been processed.\n",
      "[bwa_sai2sam_pe_core] convert to sequence coordinate...\n",
      "[infer_isize] (25, 50, 75) percentile: (133, 196, 255)\n",
      "[infer_isize] low and high boundaries: 100 and 499 for estimating avg and std\n",
      "[infer_isize] inferred external isize from 181971 pairs: 219.065 +/- 74.498\n",
      "[infer_isize] skewness: 0.680; kurtosis: 0.294; ap_prior: 1.00e-05\n",
      "[infer_isize] inferred maximum insert size: 662 (5.94 sigma)\n",
      "[bwa_sai2sam_pe_core] time elapses: 1.46 sec\n",
      "[bwa_sai2sam_pe_core] changing coordinates of 924 alignments.\n",
      "[bwa_sai2sam_pe_core] align unmapped mate...\n",
      "[bwa_paired_sw] 5898 out of 31489 Q17 singletons are mated.\n",
      "[bwa_paired_sw] 22 out of 30079 Q17 discordant pairs are fixed.\n",
      "[bwa_sai2sam_pe_core] time elapses: 0.89 sec\n",
      "[bwa_sai2sam_pe_core] refine gapped alignments... 0.25 sec\n",
      "[bwa_sai2sam_pe_core] print alignments... 0.60 sec\n",
      "[bwa_sai2sam_pe_core] 3932160 sequences have been processed.\n",
      "[bwa_sai2sam_pe_core] convert to sequence coordinate...\n",
      "[infer_isize] (25, 50, 75) percentile: (131, 195, 253)\n",
      "[infer_isize] low and high boundaries: 100 and 497 for estimating avg and std\n",
      "[infer_isize] inferred external isize from 182545 pairs: 217.849 +/- 74.041\n",
      "[infer_isize] skewness: 0.682; kurtosis: 0.300; ap_prior: 1.01e-05\n",
      "[infer_isize] inferred maximum insert size: 658 (5.94 sigma)\n",
      "[bwa_sai2sam_pe_core] time elapses: 1.37 sec\n",
      "[bwa_sai2sam_pe_core] changing coordinates of 918 alignments.\n",
      "[bwa_sai2sam_pe_core] align unmapped mate...\n",
      "[bwa_paired_sw] 6658 out of 29584 Q17 singletons are mated.\n",
      "[bwa_paired_sw] 29 out of 30844 Q17 discordant pairs are fixed.\n",
      "[bwa_sai2sam_pe_core] time elapses: 0.92 sec\n",
      "[bwa_sai2sam_pe_core] refine gapped alignments... 0.25 sec\n",
      "[bwa_sai2sam_pe_core] print alignments... 0.60 sec\n",
      "[bwa_sai2sam_pe_core] 4194304 sequences have been processed.\n",
      "[bwa_sai2sam_pe_core] convert to sequence coordinate...\n",
      "[infer_isize] (25, 50, 75) percentile: (129, 193, 251)\n",
      "[infer_isize] low and high boundaries: 100 and 495 for estimating avg and std\n",
      "[infer_isize] inferred external isize from 169583 pairs: 216.286 +/- 73.526\n",
      "[infer_isize] skewness: 0.688; kurtosis: 0.318; ap_prior: 1.00e-05\n",
      "[infer_isize] inferred maximum insert size: 653 (5.94 sigma)\n",
      "[bwa_sai2sam_pe_core] time elapses: 1.34 sec\n",
      "[bwa_sai2sam_pe_core] changing coordinates of 849 alignments.\n",
      "[bwa_sai2sam_pe_core] align unmapped mate...\n",
      "[bwa_paired_sw] 11187 out of 35982 Q17 singletons are mated.\n",
      "[bwa_paired_sw] 26 out of 29497 Q17 discordant pairs are fixed.\n",
      "[bwa_sai2sam_pe_core] time elapses: 1.11 sec\n",
      "[bwa_sai2sam_pe_core] refine gapped alignments... 0.24 sec\n",
      "[bwa_sai2sam_pe_core] print alignments... 0.61 sec\n",
      "[bwa_sai2sam_pe_core] 4456448 sequences have been processed.\n",
      "[bwa_sai2sam_pe_core] convert to sequence coordinate...\n",
      "[infer_isize] (25, 50, 75) percentile: (128, 192, 250)\n",
      "[infer_isize] low and high boundaries: 100 and 494 for estimating avg and std\n",
      "[infer_isize] inferred external isize from 141969 pairs: 215.537 +/- 72.905\n",
      "[infer_isize] skewness: 0.684; kurtosis: 0.312; ap_prior: 1.00e-05\n",
      "[infer_isize] inferred maximum insert size: 649 (5.94 sigma)\n",
      "[bwa_sai2sam_pe_core] time elapses: 1.16 sec\n",
      "[bwa_sai2sam_pe_core] changing coordinates of 697 alignments.\n",
      "[bwa_sai2sam_pe_core] align unmapped mate...\n",
      "[bwa_paired_sw] 8036 out of 56303 Q17 singletons are mated.\n",
      "[bwa_paired_sw] 24 out of 25471 Q17 discordant pairs are fixed.\n",
      "[bwa_sai2sam_pe_core] time elapses: 0.88 sec\n",
      "[bwa_sai2sam_pe_core] refine gapped alignments... 0.23 sec\n",
      "[bwa_sai2sam_pe_core] print alignments... 0.54 sec\n",
      "[bwa_sai2sam_pe_core] 4718592 sequences have been processed.\n",
      "[bwa_sai2sam_pe_core] convert to sequence coordinate...\n",
      "[infer_isize] (25, 50, 75) percentile: (133, 195, 252)\n",
      "[infer_isize] low and high boundaries: 100 and 490 for estimating avg and std\n",
      "[infer_isize] inferred external isize from 56800 pairs: 216.463 +/- 72.633\n",
      "[infer_isize] skewness: 0.654; kurtosis: 0.243; ap_prior: 1.00e-05\n",
      "[infer_isize] inferred maximum insert size: 648 (5.94 sigma)\n",
      "[bwa_sai2sam_pe_core] time elapses: 0.55 sec\n",
      "[bwa_sai2sam_pe_core] changing coordinates of 274 alignments.\n",
      "[bwa_sai2sam_pe_core] align unmapped mate...\n",
      "[bwa_paired_sw] 31817 out of 69997 Q17 singletons are mated.\n",
      "[bwa_paired_sw] 8 out of 9226 Q17 discordant pairs are fixed.\n",
      "[bwa_sai2sam_pe_core] time elapses: 1.36 sec\n",
      "[bwa_sai2sam_pe_core] refine gapped alignments... 0.15 sec\n",
      "[bwa_sai2sam_pe_core] print alignments... 0.33 sec\n",
      "[bwa_sai2sam_pe_core] 4886129 sequences have been processed.\n",
      "[main] Version: 0.7.17-r1188\n",
      "[main] CMD: /home/will/src_and_bin/Other_Software/miniconda3/envs/tnseq_avium_abx_fordistro/bin/bwa sampe input/Genome/GCF_000195955.2_ASM19595v2_genomic.fna output/tpp_out/SRR23906897.sai1 output/tpp_out/SRR23906897.sai2 output/tpp_out/SRR23906897.trimmed1 output/tpp_out/SRR23906897.genomic2\n",
      "[main] Real time: 67.439 sec; CPU: 66.480 sec\n",
      "\n",
      "[tn_preprocess] tabulating template counts and statistics for reference genome input/Genome/GCF_000195955.2_ASM19595v2_genomic.fna\n",
      "[tn_preprocess] writing output/tpp_out/SRR23906897.wig\n",
      "[tn_preprocess] writing output/tpp_out/SRR23906897.tn_stats\n",
      "[tn_preprocess] Done.\n",
      "[!] [WARNING]: Mismatches within the second line of metadata from the Wiggle files.\n"
     ]
    }
   ],
   "source": [
    "%%bash\n",
    "#SET THESE VARIABLES FOR YOUR LOCAL FILE STRUCTURE:\n",
    "PATH_TO_DATA=input/fastq #Note: Any fastq files located here will be processed for alignment with tpp\n",
    "OUT_DIR=output\n",
    "TPP_OUT_DIR=tpp_out\n",
    "REFS=input/Genome/GCF_000195955.2_ASM19595v2_genomic.fna\n",
    "GENBANK_FILE=input/Genome/GCF_000195955.2_ASM19595v2_genomic.gbff\n",
    "\n",
    "#Definition of variables for processing\n",
    "PYTHON2=$(which python2)\n",
    "BWA=$(which bwa)\n",
    "BWA_ALG=\"aln\"\n",
    "\n",
    "REPLICON_ID=\"NC_000962.3\"\n",
    "FASTQ_DIR=$PATH_TO_DATA\n",
    "\n",
    "PREFIXES_OUTFILE=$OUT_DIR/$TPP_OUT_DIR/`basename $FASTQ_DIR`_prefixes.txt\n",
    "\n",
    "# These are used for creating a CSV file\n",
    "CSV_OUTFILE=$OUT_DIR/`basename $FASTQ_DIR`.csv\n",
    "UNIQUE_FIELDS=\"locus_tag\"\n",
    "FIELDS=\"product regulatory_class bound_moiety\"\n",
    "\n",
    "#Parameter settings for tpp\n",
    "PRIMER=AACCTGTTA\n",
    "MISMATCHES=2\n",
    "WINDOW_SIZE=6\n",
    "\n",
    "###################################################################\n",
    "\n",
    "#Gunzip all fastq.gz files:\n",
    "#gunzip -k $FASTQ_DIR/*.fastq.gz\n",
    "\n",
    "#Process raw fastq files using tpp\n",
    "COUNTER=0\n",
    "INITIAL_START_TIME=$SECONDS\n",
    "for FASTQ in $FASTQ_DIR/*_1.fastq; do\n",
    "#for FASTQ in $FASTQ_DIR/Rich_6d_H+_1_R1.fastq; do\n",
    "    echo \"******** Run $COUNTER: $FASTQ ********\"\n",
    "    READS1=$FASTQ\n",
    "    READS2=${FASTQ/_1.fastq/_2.fastq}\n",
    "\n",
    "    OUTNAME=$(basename $FASTQ)\n",
    "    OUTNAME=${OUTNAME/_1.fastq/}\n",
    "    tpp -himar1 -bwa $BWA -bwa-alg $BWA_ALG -ref $REFS -replicon-ids $REPLICON_ID -reads1 $READS1 -reads2 $READS2 \\\n",
    "       -window-size $WINDOW_SIZE -primer $PRIMER -mismatches $MISMATCHES -output $OUT_DIR/$TPP_OUT_DIR/$OUTNAME &\n",
    "done\n",
    "wait\n",
    "\n",
    "echo \"Creating prefixes file with all prefixes from all runs...\"\n",
    "basename -a $OUT_DIR/$TPP_OUT_DIR/*.wig | rev | cut -c-4 --complement | rev | uniq > $PREFIXES_OUTFILE\n",
    "echo \"Created '$PREFIXES_OUTFILE'.\"\n",
    "echo \"\"\n",
    "echo \"Creating CSV file with all samples processed by TPP...\"\n",
    "$PYTHON2 scripts/wig_gb_to_csv.py -l $PREFIXES_OUTFILE -g $GENBANK_FILE -u $UNIQUE_FIELDS -f $FIELDS -o $CSV_OUTFILE\n",
    "echo \"Created '$CSV_OUTFILE'.\"\n",
    "echo \"\"\n",
    "(( TOTAL_RUN_TIME = SECONDS - INITIAL_START_TIME )) \n",
    "echo \"********** TPP driver script finished in a total of $TOTAL_RUN_TIME seconds **********\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "JT test run time: 2.134294033050537\n",
      "JT test run time: 134.73224663734436\n",
      "JT test run time: 1.8231902122497559\n",
      "JT test run time: 133.4129159450531\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/will/Professional/JHU/Research/baderlab/Drafts/2022-08-06_Mtb_Hypersusceptibility_paper/final_code/PBS_analysis/scripts/hypersus_helper.py:61: FutureWarning: In a future version of pandas, a length 1 tuple will be returned when iterating over a groupby with a grouper equal to a list of length 1. Don't supply a list with a single grouper to avoid this warning.\n",
      "  for name,g in df.groupby(['uid'],sort=False):\n",
      "/home/will/Professional/JHU/Research/baderlab/Drafts/2022-08-06_Mtb_Hypersusceptibility_paper/final_code/PBS_analysis/scripts/hypersus_helper.py:61: FutureWarning: In a future version of pandas, a length 1 tuple will be returned when iterating over a groupby with a grouper equal to a list of length 1. Don't supply a list with a single grouper to avoid this warning.\n",
      "  for name,g in df.groupby(['uid'],sort=False):\n",
      "/home/will/Professional/JHU/Research/baderlab/Drafts/2022-08-06_Mtb_Hypersusceptibility_paper/final_code/PBS_analysis/scripts/hypersus_helper.py:61: FutureWarning: In a future version of pandas, a length 1 tuple will be returned when iterating over a groupby with a grouper equal to a list of length 1. Don't supply a list with a single grouper to avoid this warning.\n",
      "  for name,g in df.groupby(['uid'],sort=False):\n",
      "/home/will/Professional/JHU/Research/baderlab/Drafts/2022-08-06_Mtb_Hypersusceptibility_paper/final_code/PBS_analysis/scripts/hypersus_helper.py:61: FutureWarning: In a future version of pandas, a length 1 tuple will be returned when iterating over a groupby with a grouper equal to a list of length 1. Don't supply a list with a single grouper to avoid this warning.\n",
      "  for name,g in df.groupby(['uid'],sort=False):\n"
     ]
    }
   ],
   "source": [
    "%%bash\n",
    "#Run analysis script (depends on the file output/input.csv in order to run)\n",
    "\n",
    "python3 scripts/hypersus_analysis.py "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/will/src_and_bin/Other_Software/miniconda3/envs/tnseq_avium_abx_fordistro/lib/python2.7/site-packages/ipykernel_launcher.py:42: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy\n",
      "/home/will/src_and_bin/Other_Software/miniconda3/envs/tnseq_avium_abx_fordistro/lib/python2.7/site-packages/ipykernel_launcher.py:43: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy\n"
     ]
    }
   ],
   "source": [
    "#Reorganize data for TableS1\n",
    "infold = 'output/hypersus_analysis/'\n",
    "outfold = 'output/hypersus_analysis/'\n",
    "\n",
    "lfc_thresh = 0.5\n",
    "padj_thresh = 0.05\n",
    "\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "from math import log10, floor\n",
    "\n",
    "def round_sig(x, sig=2):\n",
    "    #Implement significant figures in python\n",
    "    if np.isnan(x):\n",
    "        return(np.nan)\n",
    "    return round(x, sig-int(floor(log10(abs(x))))-1)\n",
    "\n",
    "def make_tables1(infilename12h, infilename48h, outfilename, treatment, drug_dose_ls):\n",
    "    #Merges tables output from analysis and changes formatting for disribution (TableS1)\n",
    "    \n",
    "    abx12_df = pd.read_csv(infilename12h).rename({'pval-adj (BH)':'padj'},axis='columns')\n",
    "    abx48_df = pd.read_csv(infilename48h).rename({'pval-adj (BH)':'padj'},axis='columns')\n",
    "\n",
    "    abx12_df[treatment+['pval','padj']] = abx12_df[treatment+['pval','padj']].applymap(round_sig)\n",
    "    abx48_df[treatment+['pval','padj']] = abx48_df[treatment+['pval','padj']].applymap(round_sig)\n",
    "\n",
    "    merged_df = abx12_df.merge(abx48_df, how='inner',on=['uid','product'],suffixes=('_7d','_14d'))\n",
    "    merged_df['Prediction'] = 'Not Significant'\n",
    "    hypersus_bool = ((merged_df['padj_7d'] < padj_thresh) & \\\n",
    "                (merged_df[treatment[-1]+'_7d'] < -lfc_thresh)) & \\\n",
    "                ((merged_df['padj_14d'] < padj_thresh) & \\\n",
    "                (merged_df[treatment[-1]+'_14d'] < -lfc_thresh))\n",
    "    hypertol_bool = ((merged_df['padj_7d'] < padj_thresh) & \\\n",
    "                (merged_df[treatment[-1]+'_7d'] > lfc_thresh)) & \\\n",
    "                ((merged_df['padj_14d'] < padj_thresh) & \\\n",
    "                (merged_df[treatment[-1]+'_14d'] > lfc_thresh))\n",
    "\n",
    "    if (len(treatment) == 1):\n",
    "        merged_df = merged_df.rename({treatment[0]+'_7d':drug_dose_ls[0]+' - LFC_7d', \\\n",
    "            treatment[0]+'_14d':drug_dose_ls[0]+' - LFC_14d'}, axis=1)\n",
    "\n",
    "    elif (len(treatment) ==2):\n",
    "        merged_df = merged_df.rename({treatment[0]+'_7d':drug_dose_ls[0]+' - LFC_7d', \\\n",
    "            treatment[1]+'_7d':drug_dose_ls[1]+' - LFC_7d', \\\n",
    "            treatment[0]+'_14d':drug_dose_ls[0]+' - LFC_14d', \\\n",
    "            treatment[1]+'_14d':drug_dose_ls[1]+' - LFC_14d'}, axis=1).drop(['R+_7d','R+_14d'],axis=1)\n",
    "\n",
    "    else:\n",
    "        exit(\"Number of doses should be either 1 or 2\")\n",
    "    merged_df = merged_df.rename({'uid':'Genomic Feature','product':'Genbank Annotation'}, axis=1)\n",
    "\n",
    "    merged_df['Prediction'][hypersus_bool]='Hypersusceptible'\n",
    "    merged_df['Prediction'][hypertol_bool]='Hypertolerant'\n",
    "\n",
    "    merged_df.to_csv(outfilename,index=False)\n",
    "    \n",
    "treatment = ['R++','R+++']\n",
    "drug_dose_ls = ['Rifampin (0.4ug/mL)','Rifampin (4.0ug/mL)']\n",
    "make_tables1(infold+'summary_7d_RMP.csv', infold+'summary_14d_RMP.csv', outfold+'TableS3_RMP.csv', treatment, drug_dose_ls)\n",
    "    \n",
    "treatment = ['H+++']\n",
    "drug_dose_ls = ['Isoniazid (1.0ug/mL)']\n",
    "make_tables1(infold+'summary_7d_INH.csv', infold+'summary_14d_INH.csv', outfold+'TableS4_INH.csv', treatment, drug_dose_ls)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 2
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython2",
   "version": "2.7.15"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
